js-packages/@prestojs/rest/src/Endpoint.ts

```diff
@@ -669,49 +728,54 @@ export default class Endpoint<ReturnT = any> {
                     .then(async res => {
                         returnVal.response = res;
                         if (!res.ok) {
-                            throw new ApiError(
-                                res.status,
-                                res.statusText,
-                                await this.decodeBody(res)
-                            );
+                            returnVal.decodedBody = await this.decodeBody(res);
+                            throw new ApiError(res.status, res.statusText, returnVal.decodedBody);
                         }
                         return res;
                     })
-                    .then(res => this.decodeBody(res))
-                    .then(data => {
-                        returnVal.decodedBody = data;
-                        if (paginator) {
-                            // If we have a paginator update its state based on response
-                            // This runs for all requests but should return false if response
-                            // is not paginated.
-                            const paginationState = cls.defaultConfig.getPaginationState(
-                                paginator,
-                                returnVal
-                            );
-                            if (paginationState) {
-                                paginator.setResponse(paginationState);
-                                data = paginationState.results;
-                            } else {
-                                // TODO: If you specify a paginator and the response is not paginated should
-                                // we warn? I think yes as I can't think of a reason why not but may need to
-                                // revisit this if we find a usecase
-                                console.warn(
-                                    'A paginator was defined but the response was not paginated. Paginator state has not been updated.'
-                                );
-                            }
-                        }
-                        return this.transformResponseBody ? this.transformResponseBody(data) : data;
-                    });
+                    .then(async res => (returnVal.decodedBody = await this.decodeBody(res)));
             const executeWithMiddleware = (): Promise<ReturnT> => {
                 const middleware = [...this.middleware];
                 let lastMiddleware;
+                const middlewareContext: MiddlewareContext<ReturnT> = {
+                    execute: executeWithMiddleware,
+                    executeOptions: options,
+                    requestInit,
+                    endpoint: this,
+                    get url(): string {
```

Comment:
See comments earlier: if you have a pre/post fetch middleware contexts then typescript should be able to catch these errors

Reply:
as above not possible

and still really want to catch them at runtime if not using typescript